### PR TITLE
Implementation of the subscriptions model

### PR DESCRIPTION
**Pull request by emmdim**
Adds plans and subscriptions and the corresponding stripe webhook and transaction checker.

**Plans** are the available SaaS plans, that are also correlated with the stripe plans.
**Subscriptions** is the relation between an organization  and and plan.
The **Stripe Webhook** is responsible for receiving the stripe subscription updates for each organization.
The **transaction checker** is implemented in the new `subscriptions` package, that sets the base for restricting the access to features that are not allowed by the subscription

Some technical details:
* The desciption of the subscriptions' features (as well as their stripe ids) is define in the `assets/plans.json` file.
* The above file should contain a default plan (`default:true`) that is assigned to all new organizations, containing the minimum features
* The organization struct already includes some basic subsription information
* The detailed information about an organization subscription is provided by the `GET /organizations/{address}/subscription` call made by an organization admin

View the following docs sections for more info requests :
https://github.com/vocdoni/saas-backend/blob/c39e5263dfd9fb5ecb997725d5451630b8bea737/api/docs.md?plain=1#L677
https://github.com/vocdoni/saas-backend/blob/c39e5263dfd9fb5ecb997725d5451630b8bea737/api/docs.md?plain=1#L825

### CODE REVIEW

File: db/types.go [hunk 1]
```diff
@@ -40,24 +40,59 @@ type OrganizationMember struct {
...
+	Subscription    OrganizationSubscription `json:"subscription" bson:"subscription"`
+}
+
+type OrganizationLimits struct {
+	Memberships   int `bson:"memberships"`
+	SubOrgs       int `bson:"subOrgs"`
+	MaxCensusSize int `bson:"maxCensusSize"`
+}
+
+type VotingTypes struct {
```
[the remark this answers is not in the document]
Comment (emmdim) [reply db/types.go:72]: I am not sure what you mean. 
It is part of the `Subscritpion` collection:
https://github.com/vocdoni/saas-backend/blob/7619754d1e4a4f5de0c56bd693c371467928d3ad/db/types.go#L89

[the remark this answers is not in the document]
Comment (lucasmenendez) [reply db/types.go:72]: after today meeting i understand why these attributes are not included in the `Subscription` object

File: api/stripe.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+	// Unmarshal the event data into an appropriate struct depending on its Type
+	switch event.Type {
+	case "customer.subscription.created":
+		customer, subscription, err := a.stripe.GetInfoFromEvent(*event)
+		if err != nil {
+			log.Warnw("Stripe Webhook: Error getting info from event: %v\n", err)
+			w.WriteHeader(http.StatusBadRequest)
+			return
+		}
+		email := customer.Email
```
Comment (lucasmenendez) [inline api/stripe.go:44]: why dont use `customer.email` directly?

File: api/stripe.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+	"go.vocdoni.io/dvote/log"
+)
+
+// handleWebhook handles the incoming webhook event from Stripe.
+// It takes the API data and signature as input parameters and returns the session ID and an error (if any).
+// The request body and Stripe-Signature header are passed to ConstructEvent, along with the webhook signing key.
+// If the event type is "customer.subscription.created", it unmarshals the event data into a CheckoutSession struct
+// and returns the session ID. Otherwise, it returns an empty string.
+func (a *API) handleWebhook(w http.ResponseWriter, r *http.Request) {
+	const MaxBodyBytes = int64(65536)
```
Comment (lucasmenendez) [inline api/stripe.go:18]: Why is the maximum required?

File: stripe/stripe.go [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+func (s *StripeClient) GetInfoFromEvent(event stripe.Event) (*stripe.Customer, *stripe.Subscription, error) {
+	var subscription stripe.Subscription
+	err := json.Unmarshal(event.Data.Raw, &subscription)
+	if err != nil {
+		log.Warnf("Error parsing webhook JSON: %s\n", err.Error())
+		return nil, nil, err
+	}
+	log.Warnf("Subscription created for %s.", subscription.ID)
+	log.Warnf("Subscription plan for %s.", subscription.Items.Data[0].Plan.ID)
```
Comment (lucasmenendez) [inline stripe/stripe.go:48]: `subscription.Items.Data` can be nil here, even the unmarshal success

File: db/helpers.go [hunk 5]
```diff
@@ -180,3 +215,34 @@ func dynamicUpdateDocument(item interface{}, alwaysUpdateTags []string) (bson.M,
 	}
 	return bson.M{"$set": update}, nil
 }
+
+// readSubscriptionJSON reads a JSON file with an array of subscritpions
+// and return it as a Subscription array
+func readSubscriptionJSON(subscriptionsFile string) ([]*Subscription, error) {
+	if subscriptionsFile == "" {
+		subscriptionsFile = "subscriptions.json"
```
Comment (lucasmenendez) [inline db/helpers.go:223]: this probably should be a constant

Comment (lucasmenendez) [reply db/helpers.go:223]: and probably should be configurable via envvars or cli args

**Review by lucasmenendez — CHANGES_REQUESTED** [review]

File: stripe/stripe.go [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+	"github.com/stripe/stripe-go/v80/customer"
+	"github.com/stripe/stripe-go/v80/webhook"
+	"go.vocdoni.io/dvote/log"
+)
+
+type StripeClient struct {
+	webhookSecret string
+}
+
+func New(apiSecret, webhookSecret string) (*StripeClient, error) {
```
[the remark this answers is not in the document]
Comment (emmdim) [reply stripe/stripe.go:16]: Got me. It was a previous check that got removed

File: api/stripe.go (hunk 3 above)
Comment (emmdim) [reply api/stripe.go:18]: It is defined like that from stripe

File: db/validations.go [hunk 7]
```diff
@@ -29,3 +31,103 @@ var usersCollectionValidator = bson.M{
...
+			},
+			"name": bson.M{
+				"bsonType":    "string",
+				"description": "the name of the subscription plan must be a string and is required",
+			},
+			"stripeID": bson.M{
+				"bsonType":    "string",
+				"description": "the corresponding plan ID must be a string and is required",
+			},
+			// 	"organization": bson.M{
```
[the remark this answers is not in the document]
Comment (emmdim) [reply db/validations.go:71]: Will come back to it when  the subscirptions features options are more solid

File: db/helpers.go (hunk 5 above)
Comment (emmdim) [reply db/helpers.go]: @lucasmenendez  can you review that the weay the env var is handled is OK with you?

**Review by lucasmenendez — APPROVED** [review]

**Review by lucasmenendez — CHANGES_REQUESTED** [review]
there are some misspellings to fix

File: db/helpers.go [hunk 8]
```diff
@@ -43,6 +52,20 @@ func (ms *MongoStorage) initCollections(database string) error {
 					return nil, fmt.Errorf("failed to update collection validator: %w", err)
 				}
 			}
+			if name == "subscriptions" {
+				// clear subscriptions collection and update the DB with the new ones
+				if _, err := ms.client.Database(database).Collection(name).DeleteMany(ctx, bson.D{}); err != nil {
```
[the remark this answers is not in the document]
Comment (emmdim) [reply db/helpers.go:56]: But If I move it outside the "if alreadyCreated" check i would have to do anyway an exta check `if name == "subscriptions"`

File: api/transaction.go [hunk 9]
```diff
@@ -110,13 +111,19 @@ func (a *API) signTxHandler(w http.ResponseWriter, r *http.Request) {
 		case *models.Tx_NewProcess:
 			txNewProcess := tx.GetNewProcess()
 			// check the tx fields
-			if txNewProcess == nil || txNewProcess.Process == nil || txNewProcess.Nonce == 0 {
+			if txNewProcess == nil || txNewProcess.Process == nil {
+				// if txNewProcess == nil || txNewProcess.Process == nil || txNewProcess.Nonce == 0 {
```
Comment (lucasmenendez) [inline api/transaction.go:115]: :(

File: subscriptions/subscriptions.go [hunk 10]
```diff
@@ -0,0 +1,45 @@
+package subscriptions
```
Comment (lucasmenendez) [inline subscriptions/subscriptions.go:1]: Please before finish the PR, add comments to public functions and structs

**Review by lucasmenendez — APPROVED** [review]

File: db/organizations.go [hunk 11]
```diff
@@ -54,6 +54,37 @@ func (ms *MongoStorage) Organization(address string, parent bool) (*Organization
...
+// it should be returned, it returns the specific error. If other errors occur,
+// it returns the error.
+func (ms *MongoStorage) OrganizationByCreatorEmail(email string, parent bool) (*Organization, *Organization, error) {
+	ms.keysLock.RLock()
+	defer ms.keysLock.RUnlock()
+	// create a context with a timeout
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	// find the organization in the database
+	result := ms.organizations.FindOne(ctx, bson.M{"creator": email})
```
Comment (lucasmenendez) [inline db/organizations.go:69]: Should the `creator` attribute have an index in the database?

File: db/organizations.go [hunk 12]
```diff
@@ -155,3 +186,43 @@ func (ms *MongoStorage) OrganizationsMembers(address string) ([]User, error) {
 	}
 	return users, nil
 }
+
+// addSubscriptionToOrganization internal method adds the subscription to the organiation with
```
Comment (lucasmenendez) [inline db/organizations.go:190]: the comment does not match with the current method name and implementation

File: db/organizations.go [hunk 13]
```diff
@@ -155,3 +186,43 @@ func (ms *MongoStorage) OrganizationsMembers(address string) ([]User, error) {
...
+	filter := bson.M{"_id": address}
+	updateDoc := bson.M{"$set": bson.M{"subscription": orgSubscription}}
+	// update the organization in the database
+	if _, err := ms.organizations.UpdateOne(ctx, filter, updateDoc); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Implement update subscription method and ensuere that only members
```
Comment (lucasmenendez) [inline db/organizations.go:212]: The function comments in go should start with the name of the function itself: `UpdateOrganizationSubscritpion function implements...`

File: db/organizations.go [hunk 14]
```diff
@@ -155,3 +186,43 @@ func (ms *MongoStorage) OrganizationsMembers(address string) ([]User, error) {
...
+// with adequate role can update the subscription
+func (ms *MongoStorage) UpdateOrganizationSubscritpion(address string, orgSubscription *OrganizationSubscription) error {
```
Comment (lucasmenendez) [inline db/organizations.go:214]: This method is not used anywhere. Will it be needed in the future?

File: db/types.go [hunk 15]
```diff
@@ -40,21 +42,66 @@ type OrganizationMember struct {
...
+	Subdomain       string                   `json:"subdomain" bson:"subdomain"`
+	Country         string                   `json:"country" bson:"country"`
+	Timezone        string                   `json:"timezone" bson:"timezone"`
+	Active          bool                     `json:"active" bson:"active"`
+	TokensPurchased uint64                   `json:"tokensPurchased" bson:"tokensPurchased"`
+	TokensRemaining uint64                   `json:"tokensRemaining" bson:"tokensRemaining"`
+	Parent          string                   `json:"parent" bson:"parent"`
+	Subscription    OrganizationSubscription `json:"subscription" bson:"subscription"`
+	Counters        OrganizationCounters     `json:"counters" bson:"counters"`
+}
```
Comment (lucasmenendez) [inline db/types.go:62]: Please add a linebreak between structs

File: db/validations.go [hunk 16]
```diff
@@ -23,7 +24,7 @@ var usersCollectionValidator = bson.M{
 			"email": bson.M{
 				"bsonType":    "string",
 				"description": "must be an email and is required",
-				"pattern":     internal.EmailRegexTemplate,
+				"pattern":     `^[\w.\+\.\-]+@([\w\-]+\.)+[\w]{2,}$`,
```
Comment (lucasmenendez) [inline db/validations.go:27]: Why not to use the same regex everywhere? That is why this variable exists.

File: stripe/stripe.go [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+		webhookSecret: webhookSecret,
+	}
+}
+
+// DecodeEvent decodes a Stripe webhook event from the given payload and signature header.
+func (s *StripeClient) DecodeEvent(payload []byte, signatureHeader string) (*stripe.Event, error) {
+	event := stripe.Event{}
+
+	if err := json.Unmarshal(payload, &event); err != nil {
+		log.Warnw("Stripe Webhook: error while parsing basic request. %s\n", err.Error())
```
Comment (lucasmenendez) [inline stripe/stripe.go:32]: Same, please do not capitalize error messages.

File: api/stripe.go [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+	signatureHeader := r.Header.Get("Stripe-Signature")
+	event, err := a.stripe.DecodeEvent(payload, signatureHeader)
+	if err != nil {
+		log.Warnw("Stripe Webhook: Error decoding event: %v\n", err)
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+	// Unmarshal the event data into an appropriate struct depending on its Type
+	switch event.Type {
+	case "customer.subscription.created":
```
[the remark this answers is not in the document]
Comment (emmdim) [reply api/stripe.go:37]: Correct, will start to add cases  with the next PR :-)

File: db/organizations.go (hunk 14 above)
Comment (emmdim) [reply db/organizations.go:214]: This organization will be necessary wonce the stripe webhook handling is complete

File: db/organizations.go (hunk 12 above)
Comment (emmdim) [reply db/organizations.go:190]: Updated comment and function renamed to SetOrganizationSubscritption.

File: db/organizations.go (hunk 13 above)
Comment (emmdim) [reply db/organizations.go:212]: Removed funciont as it was redudant with `SetOrganizationSubscritption`

File: api/types.go [hunk 19]
```diff
@@ -147,6 +149,11 @@ type MessageSignature struct {
 	Signature types.HexBytes `json:"signature,omitempty"`
 }
 
+type StripeConfig struct {
```
[the remark this answers is not in the document]
Comment (emmdim) [reply api/types.go:152]: Correct but it was redundant and got deleted.

**Review by lucasmenendez — APPROVED** [review]

**Pull request merged**